### PR TITLE
[Rest API Compatibility] Typed endpoints for Put and Get Mapping and get field mappings

### PR DESCRIPTION
**Pull request by pgomulka**
Implements a V7 compatible typed endpoints for REST put and get mapping endpoints. Also for Get Field Mappings endpoints.
retrofits the REST layer change removed in https://github.com/elastic/elasticsearch/pull/41676

relates main meta issue https://github.com/elastic/elasticsearch/issues/51816
relates types removal issue https://github.com/elastic/elasticsearch/issues/54160

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestPutMappingAction.java [hunk 1]
```diff
@@ -48,15 +66,47 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
...
+            }
+            final String type = request.param("type");
+            if (includeTypeName == false &&
+                (type != null || isMappingSourceTyped(MapperService.SINGLE_MAPPING_NAME, sourceAsMap))) {
+                throw new IllegalArgumentException("Types cannot be provided in put mapping requests, unless " +
+                    "the include_type_name parameter is set to true.");
+            }
+            // TODO what about the type in the source? it should be skipped?
+            //  (different than in RestCreateIndexAction which expects the _doc)
+            Map<String, Object> mappingSource = prepareV7Mappings(includeTypeName, sourceAsMap);
```
Comment (pgomulka) [inline server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestPutMappingAction.java:83]: my point here is that when a put mapping request on endpoint `PUT {index}/_mapping`
with body
```
PUT my_index 
{
  "mappings": {
    "MY_TYPE": {
      "properties": {
        "name": {
          "properties": {
            "first": {
              "type": "text"
            }
          }
        },
        "user_id": {
          "type": "keyword"
        }
      }
    }
  }
}
```

could have MY_TYPE which I simply omit. The `putMappingRequest.source` will be provided with a map which starts with `properties`

**Comment by pgomulka** [comment]
once again, this PR includes a testing changes which will be removed later. I will merge them on a separate PR. Feel free to ignore them.

**Review by romseygeek — APPROVED** [review]
I can't comment on the test rewrite code but LGTM from the mappings point of view.

File: buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/transform/match/ReplaceLength.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.gradle.test.rest.transform.match;
+
+import org.gradle.api.tasks.Internal;
+
+/**
+ * A transformation to replace the value of a match. For example, change from "match":{"_type": "foo"} to "match":{"_type": "bar"}
```
Comment (joegallo) [inline buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/transform/match/ReplaceLength.java:14]: I think the docstring here is not accurate.

Comment (pgomulka) [reply buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/transform/match/ReplaceLength.java:14]: yes, I submited a separate PR for testing part here https://github.com/elastic/elasticsearch/pull/72156/

**Pull request merged**